---
title: "Lab meeting"
author: "Bob Week"
date: "03/10/21"
output: 
  revealjs::revealjs_presentation:
    theme: night
    transition: fade
    incremental: true
    self_contained: true
    center: true
editor_options: 
  chunk_output_type: console
---

## The Signature of Coevolution<br>in Continuous Space

![https://doi.org/10.1016/B978-0-12-800049-6.00188-8](/home/bb/Projects/The Genomic Signature of Coevolution in Continuous Space/Lab Mtg/snewt.jpg)

## Motivation

- Space has long been thought to play a crucial role in the coevolutionary process

  - Local adaptation and gene-flow may lead to patterns of trait mismatch and maladaptation
  
- Past theoretical work has largely ignored the relationship between coevolution and IBD

  - This is particularly troublesome for methods seeking to identify coevolving loci
  
- Here we outline three projects that attempt to answer the following questions

## Questions

- At what spatial scales do interspecific coevolutionary patterns emerge?

  - The spatial scale of interspecific trait/genotype cross-correlations?

- How does this relate to spatial scales of intraspecific local adaptation?

  - How can we measure local adaptation?

- How do these spatial scales and their relationships depend on dispersal distances and environmental heterogeneity?

- Can we disentangle the signature of coevolutionary adaptation from isolation by distance?
  
## Outline

- We focus on host-parasite coevolution

- Three projects:
  
  1. Phenotypic
  
  2. Two-locus
  
  3. SLiM

## Phenotypic

This initial project seeks to answer questions relating to spatial scales of coevolution, local adaptation and environmental heterogeneity using a quantitative-genetic model. By focusing on phenotypic evolution, we may obtain a more broadscale understanding of these phenomena before diving into genetic/genomic details.

## Two-locus

This project injects further genetic details by modeling coevolution mediated by two loci in each species. This allows comparisons of intraspecific linkage-disequilibrium to interspecific linkage-disequilibrium, thereby creating the opportunity to distinguish between interspecific linkage due to IBD and coevolution.

## SLiM

This project injects even further genetic details using genome-wide slimulations. Using these individual-based, genome-wide slimulations, we can test conclusions drawn from the simpler, but less realistic approaches taken above to see what we can expect to hold in the wild.

# Phenotypic

## Outline of Model

- Species $H$ and $P$

- Fitness $m_H,m_P$ determined by quantitative traits $z_H,z_P$

- Assuming individuals encounter each other at random, mean fitness $\bar m_H,\bar m_P$ calculated by averaging across $z_H,z_P$

- Mean trait dynamics given by

  - $\frac{d}{dt}\bar z_H=G_H\beta_H+\xi_H$
  - $\frac{d}{dt}\bar z_P=G_P\beta_P+\xi_P$

  - $G_H,G_P$ are additive genetic variances
  - $\beta_H,\beta_P$ are selection gradients $\left(=\frac{\partial\bar m_H}{\partial\bar z_H},\frac{\partial\bar m_P}{\partial\bar z_P}\right)$
  - $\xi_H,\xi_P$ are random processes capturing drift

## Trait-matching

```{r, echo=FALSE, warning=FALSE, message=FALSE}
require("ggplot2")
require("gridExtra")
require('latex2exp')
require('ggdark')

offsetmatching = function(D,Bx,By,d){ # D = x - y
  Wx = exp(-Bx*(D-d)^2/2)
  Wy = exp(-By*(D+d)^2/2)
  return(c(Wx,Wy))
}

traitdifferences = function(D,Bx,By){ # D = x - y
  Wx = exp( Bx*D)
  Wy = exp(-By*D)
  return(c(Wx,Wy))
}
Ds = seq(-1,1,0.001) # range of trait differences

td = NULL # container for trait-differences mechanism
om = NULL # container for offset-matching mechanism
for(D in Ds){
  td = rbind(td,traitdifferences(D,1.5,1))
  om = rbind(om,offsetmatching(D,10,-10,0))
}
td = 3*td/max(td) # scale max fitness to 3
om[,1] = 3*om[,1]
om[,2] = 0.25*om[,2]
om = log(om)
Fitness_df = data.frame(fit=c(td[,1],td[,2],om[,1],om[,2]), 
                        
                        Species=c( rep('P',length(Ds)), rep('H',length(Ds)), rep('P',length(Ds)), rep('H',length(Ds)) ),
                        
                        Mechanism=c( rep('TD',2*length(Ds)), rep('OM',2*length(Ds)) ),
                        
                        td=c(Ds,Ds,Ds,Ds))

OM_Fitness_df = subset(Fitness_df,Mechanism=='OM')

ggplot(data = OM_Fitness_df, aes(x=td,y=fit,linetype=Species)) + 
  
  # geom_vline(xintercept = -.5,linetype='dotted') + geom_vline(xintercept = .5,linetype='dotted') + 
  
  geom_line(size=1) + theme_minimal() + xlab(TeX('z_H-z_P')) + ylab('Fitness') +
  
  annotate("text",x=0,y=-1.75,label=TeX('m_H \ = r_H \ + \ B_H(z_H-z_P)^2')) +
  
  annotate("text",x=0,y=1.5,label=TeX('m_P \ = r_P \ - \ B_P(z_H-z_P)^2')) + dark_mode()

```

## Coevolutionary Dynamics <br> (non-spatial)

```{r, eval=FALSE, echo=FALSE}
require("ggplot2")
require('diffeqr')
require('latex2exp')
require('ggdark')

de <- diffeqr::diffeq_setup(JULIA_HOME = "/home/bb/Julia/JuliaPro-1.5.2-1/Julia/bin/")

f <- function(u,p,t){
  du = c(0,0)
  du[1] = p[1] * (-p[2]*(u[2]-u[1]) - 0.0 * u[1])
  du[2] = p[3] * ( p[4]*(u[1]-u[2]) - 0.0 * u[2])
  return(du)
}

g <- function(u,p,t){
  du = c(p[1]/p[5],p[3]/p[6])
  return(du)
}

pars <- c(10,0.01,10,0.05,1000,1000)
u0 <- c(0.1,-0.1)
tspan <- c(0,100)

prob = de$SDEProblem(f, g, u0, tspan, pars)
sol = de$solve(prob)
mat <- sapply(sol$u,identity)
udf <- data.frame(Time=c(sol$t,sol$t), Trait=c(mat[1,],mat[2,]), Species=c(rep("Host",length(sol$t)),rep("Parasite",length(sol$t))))
ggplot(data=udf,aes(x=Time,y=Trait,linetype=Species)) +
  geom_line(size=1) + theme_minimal() + xlab("Trait Value") + dark_mode()

```

![Trait Dynamics](/home/bb/Projects/The Genomic Signature of Coevolution in Continuous Space/Lab Mtg/dynamics.png)

## Adding Space

<video width="640" height="480" data-autoplay src="/home/bb/Videos/host-parasite-coevolution.mp4"></video>

- Can use theory of random fields to get spatial (intraspecific) correlation and (interspecific) cross-correlation functions.

## Spatial Correlation Functions

$D_H,D_P=$ host/parasite dispersal distance

$\rho_H,\rho_P,\rho_{HP}=$ spatial (cross)-correlations

![](/home/bb/Projects/The Genomic Signature of Coevolution in Continuous Space/Phenotypic/julia/corrs.png)

- Q: What can we conclude from this figure?

## Measuring Local Adapation

- We have fitness functions $m_H(z_H,z_P),m_P(z_P,z_H)$

- So we can calculate difference in fitness for

  - local interactions (home) vs
  - interactions at a spatial lag (away)

- $\Delta_H=\mathbb E[m_H(z_H(x),z_P(x))-m_H(z_H(x),z_P(y))]$

  - $x,y=$ spatial locations

## Local Adaptation

![](/home/bb/Projects/The Genomic Signature of Coevolution in Continuous Space/Phenotypic/julia/corrs-las.png)

## TODO

- implement environmental heterogeneity

  - random field (for clumpiness)
  
  - clines (linear=easy, but should we do more?)

## People to invite

```{css, echo=FALSE}
figure {
    display: inline-block;
}
```


<figure>
  <img data-src="aitken-stadler.jpg" width="447" height="255">
  <figcaption> 
    Sally Aitken (left) local adaptation, tree genomics<br>Tanja Stadler (right) coalescent theory, phylogenetics
  </figcaption>
</figure>
