\documentclass{article}

\usepackage{arxiv}

\usepackage[utf8]{inputenc} % allow utf-8 input
\usepackage[T1]{fontenc}    % use 8-bit T1 fonts
\usepackage{lmodern}        % https://github.com/rstudio/rticles/issues/343
\usepackage{hyperref}       % hyperlinks
\usepackage{url}            % simple URL typesetting
\usepackage{booktabs}       % professional-quality tables
\usepackage{amsfonts}       % blackboard math symbols
\usepackage{nicefrac}       % compact symbols for 1/2, etc.
\usepackage{microtype}      % microtypography
\usepackage{lipsum}
\usepackage{graphicx}

\title{A Genetic Signature of Host-Parasite Coevolution in Continuous Space}

\author{
    Bob Week
   \\
    Integrative Biology \\
    Michigan State University \\
  East Lansing, MI 48824 \\
  \texttt{\href{mailto:weekrobe@msu.edu}{\nolinkurl{weekrobe@msu.edu}}} \\
   \And
    Gideon Bradburd
   \\
    Integrative Biology \\
    Michigan State University \\
  East Lansing, MI 48824 \\
  \texttt{\href{mailto:bradburd@msu.edu}{\nolinkurl{bradburd@msu.edu}}} \\
  }


% Pandoc citation processing

\usepackage{amsmath}


\begin{document}
\maketitle

\def\tightlist{}


\begin{abstract}
Here we identify a genetic signature of host-parasite coevolution in
continuous space using spatial patterns of linkage-disequilibrium.
\end{abstract}

\keywords{
    blah
   \and
    blee
   \and
    bloo
   \and
    these are optional and can be removed
  }

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

We consider a biallelic haploid two-locus model of fitness for each
species. For the host and parasite we denote the possible haplotypes
respectively by \(p_{AB}^H,p^H_{Ab},p^H_{aB},p^H_{ab}\) and
\(p_{AB}^P,p^P_{Ab},p^P_{aB},p^P_{ab}\). Similarly, gene frequencies are
denoted by \(p^X_A\) and \(p^X_B\) for species \(X=H,P\). Linkage
disequilibrium within species \(X\) can be written as
\(D_X=p^X_{AB}-p^X_Ap^X_B\) or
\(D_X=p^X_{AB}p^X_{ab}-p^X_{Ab}p^X_{aB}\). We assume that individuals
encounter each at random and that when a host individual with haploid
genotype \(i\) encounters a parasite individual with haploid genotype
\(i\), infection occurs with probability \(\alpha_{ij}\). Given an
infection occurs, we assume the host experiences a reduction in fitness
by the amount \(s_H\). Hence, per-capita growth rates of host haplotypes
are given by

\begin{subequations}
  \begin{equation}
    m^H_{AB}=-s_H(\alpha_{AB,AB}p^P_{AB}+\alpha_{AB,Ab}p^P_{Ab}+\alpha_{AB,aB}p^P_{aB}+\alpha_{AB,ab}p^P_{ab}),
  \end{equation}
  \begin{equation}
    m^H_{aB}=-s_H(\alpha_{aB,AB}p^P_{AB}+\alpha_{aB,Ab}p^P_{Ab}+\alpha_{aB,aB}p^P_{aB}+\alpha_{aB,ab}p^P_{ab}),
  \end{equation}
  \begin{equation}
    m^H_{Ab}=-s_H(\alpha_{Ab,AB}p^P_{AB}+\alpha_{Ab,Ab}p^P_{Ab}+\alpha_{Ab,aB}p^P_{aB}+\alpha_{Ab,ab}p^P_{ab}),
  \end{equation}
  \begin{equation}
    m^H_{ab}=-s_H(\alpha_{ab,AB}p^P_{AB}+\alpha_{ab,Ab}p^P_{Ab}+\alpha_{ab,aB}p^P_{aB}+\alpha_{ab,ab}p^P_{ab}).
  \end{equation}
\end{subequations}

Similarly, by assuming infection increases per-capita growth rates of
parasite haplotypes by the amount \(s_P\), we obtain

\begin{subequations}
  \begin{equation}
    m^P_{AB}=-s_H(\alpha_{AB,AB}p^H_{AB}+\alpha_{Ab,AB}p^H_{Ab}+\alpha_{aB,AB}p^H_{aB}+\alpha_{ab,AB}p^H_{ab}),
  \end{equation}
  \begin{equation}
    m^P_{AB}=-s_H(\alpha_{AB,aB}p^H_{AB}+\alpha_{Ab,aB}p^H_{Ab}+\alpha_{aB,aB}p^H_{aB}+\alpha_{ab,aB}p^H_{ab}),
  \end{equation}
  \begin{equation}
    m^P_{Ab}=-s_H(\alpha_{AB,Ab}p^H_{AB}+\alpha_{Ab,Ab}p^H_{Ab}+\alpha_{aB,Ab}p^H_{aB}+\alpha_{ab,Ab}p^H_{ab}),
  \end{equation}
  \begin{equation}
    m^P_{ab}=-s_H(\alpha_{AB,ab}p^H_{AB}+\alpha_{Ab,ab}p^H_{Ab}+\alpha_{aB,ab}p^H_{aB}+\alpha_{ab,ab}p^H_{ab}).
  \end{equation}
\end{subequations}

The per-capita population growth rate of species \(X\) can then be
written as
\(\bar m_X=m^X_{AB}p^X_{AB}+m^X_{aB}p^X_{aB}+m^X_{Ab}p^X_{Ab}+m^X_{ab}p^X_{ab}\).
Assuming clonal reproduction, we can write down the evolution of haploid
genotype frequency \(p^X_i\) as

\begin{equation}
  \frac{dp^X_i}{dt}=(m^X_i-\bar m_X)p^X_i.
\end{equation}

However, since our interests are in sexually reproducing diploid
organisms we need to extend this model.

\hypertarget{approximation}{%
\section{Approximation}\label{approximation}}

Modeling IBD in 2D space will be challenging since the SPDE approach
breaks down here. One option is simulate the rescaled process that
approximates the superprocess solution. Another is to assume allele
frequencies are sufficiently close to 0.5 to remove the multiplicative
noise term. However, for just two loci this assumption may not hold (for
a polygenic trait it may\ldots).

\bibliographystyle{unsrt}
\bibliography{references.bib}


\end{document}
